pkg/primitives/vm/decompress_kernel.go

```diff
@@ -0,0 +1,219 @@
+	f, err := elf.Open(KernelImagePath)
+	if err != nil {
+		return err
+	}
+	f.Close()
+	return nil
+}
+
+type algoOptions struct {
+	decompressFunc func(kernelStream io.Reader) (io.Reader, error)
+	headerBytes    []byte
+	name           string
+}
+
+func decompressData(data []byte, tmpFile *os.File, o algoOptions) error {
+	var headerIndex int
+	var errs error
+
+	headerCount := bytes.Count(data, o.headerBytes)
+	if headerCount == 0 {
+		return fmt.Errorf("%s: couldn't find the compression algorithm header", o.name)
+	}
+
+	for i := 0; i < headerCount; i++ {
+		headerIndex += bytes.Index(data[headerIndex:], o.headerBytes)
+
+		headerData := data[headerIndex:]
+
+		// to ignore the current header in the next iteration
+		headerIndex += len(o.headerBytes)
+
+		r, err := o.decompressFunc(bytes.NewBuffer(headerData))
+		if err != nil {
+			errs = multierror.Append(errs, err)
+			continue
+		}
+
+		// close zstd reader after write
+		if reader, ok := r.(*zstd.Decoder); ok {
+			defer reader.Close()
+		}
+
+		_, err = io.Copy(tmpFile, r)
+		if err != nil && err != bzip2.StructuralError("bad magic value in continuation file") && err != zstd.ErrMagicMismatch {
+			errs = multierror.Append(errs, err)
+			continue
+		}
+
+		err = isValidELFKernel(tmpFile.Name())
+		if err == nil {
+			return nil
+		}
+
+		errs = multierror.Append(errs, err)
+	}
+
+	return errs
+}
+
+func gUnzip(kernelStream io.Reader) (io.Reader, error) {
+	r, err := gzip.NewReader(kernelStream)
+	if err != nil {
+		return nil, err
+	}
+
+	r.Multistream(false)
+	// TODO: how it read after closing !!!!
+	defer r.Close()
+
+	return r, nil
+}
+
+func unXZ(kernelStream io.Reader) (io.Reader, error) {
+	r, err := xz.NewReader(kernelStream, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	r.Multistream(false)
+	return r, nil
+}
+
+func bUnzip2(kernelStream io.Reader) (io.Reader, error) {
+	return bzip2.NewReader(kernelStream), nil
+}
+
+func unlzma(kernelStream io.Reader) (io.Reader, error) {
+	return lzma.NewReader(kernelStream)
+}
+
+// func lZop(kernelStream io.Reader) (io.Reader, error) {
+// 	r, err := lzo.NewReader(kernelStream)
+// 	if err != nil {
+// 		return nil, err
+// 	}
+// 	defer r.Close()
+// 	return r, nil
+// }
+
+func lZ4(kernelStream io.Reader) (io.Reader, error) {
+	return lz4.NewReader(kernelStream), nil
+}
+
+func unZstd(kernelStream io.Reader) (io.Reader, error) {
+	r, err := zstd.NewReader(kernelStream)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func tryDecompressKernel(KernelImagePath string) error {
+	if len(strings.TrimSpace(KernelImagePath)) == 0 {
+		return fmt.Errorf("kernel image is required")
+	}
+
+	// if kernel is already an elf (uncompressed)
+	if err := isValidELFKernel(KernelImagePath); err == nil {
+		return nil
+	}
+
+	// Prepare temp files:
+	tmpFile, err := os.CreateTemp("/tmp", "vmlinux-")
+	if err != nil {
+		return err
+	}
+	defer tmpFile.Close()
+	defer os.Remove(tmpFile.Name())
```

Comment:
The close has to happen first then deleting right ?
so this has to be in the opposite order. 

Or better 

```suggestion
        defer func() {
            tmpFile.Close()
            or.Remove(tmpFile.Name())
        }
```